app/controllers/spree/paypal_controller.rb

```diff
@@ -171,11 +134,11 @@ def provider
       payment_method.provider
     end
 
-    def payment_details(items)
-      item_sum = items.sum { |i| i[:Quantity] * i[:Amount][:value] }
+    def payment_details(order)
+      items = PaypalItemsBuilder.new(order).call
```

Comment:
I think I'm leaning towards preferring Concerns more these days (depending on the situation). I spoke to a Rails dev about an interesting pattern a while ago, where you extract logic into Jobs instead of Services. If you want to run the Job inline, you can just call `perform_now` (and it skips the queue and behaves exactly like a Service), but you have the option at any time of calling the same Job asynchronously with `perform_later`...

Reply:
I think `concerns` is a controversial topic. It comes in handy sometimes but as with things they need to be used properly. I've seen them being abused a lot. 

A quick googling yielded https://www.codewithjason.com/used-intelligently-rails-concerns-great/ and skimming through it, it seems it covers the topic rather well.